### ISSUE TITLE
Add cat-log command to view logs from the command-line - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1808]
_Merges catlog -> master_

Closes #930 

- [x] workflow logs
- [x] job logs
- [x] write tests
- [x] write docs

**Comment by kinow** [comment]
marked the checklist item **job logs** as completed

**Comment by kinow** [comment]
added 3 commits

<ul><li>a6909c1d - 1 commit from branch <code>master</code></li><li>c0532a71 - [cat-log] Add logic to view workflow file</li><li>c045632a - [cat-log] Add logic to view job file</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12580&start_sha=377bad491fc36c6c188b5c8775c8a47a1ae0d583)

**Comment by kinow** [comment]
added 18 commits

<ul><li>c045632a...94788c6c - 16 commits from branch <code>master</code></li><li>df7b16a5 - [cat-log] Add logic to view workflow file</li><li>ee9e4da7 - [cat-log] Add logic to view job file</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12629&start_sha=c045632ad22913502c882c2a1ca72c7b05ee5eb9)

**Comment by kinow** [comment]
added 1 commit

<ul><li>8df933ce - Add cat-log for workflow statuses too</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12637&start_sha=ee9e4da72d0041377e710b6274f22881411793c3)

**Comment by kinow** [comment]
added 6 commits

<ul><li>8df933ce...aace6b4c - 2 commits from branch <code>master</code></li><li>5bb7157a - [cat-log] Add logic to view workflow file</li><li>9300f6c4 - [cat-log] Add logic to view job file</li><li>b86f4bdb - Add cat-log for workflow statuses too</li><li>54ff1a07 - WIP docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12645&start_sha=8df933ceab48ea61c79bdb57a2c819d20f9bc70a)

**Comment by kinow** [comment]
added 43 commits

<ul><li>54ff1a07...4f7750f3 - 39 commits from branch <code>master</code></li><li>84e15a6c - [cat-log] Add logic to view workflow file</li><li>705fb535 - [cat-log] Add logic to view job file</li><li>c9d8e050 - Add cat-log for workflow statuses too</li><li>55e34321 - WIP docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12936&start_sha=54ff1a076e817704590d70bc73be51c45e695441)

**Comment by kinow** [comment]
added 6 commits

<ul><li>55e34321...365c9919 - 2 commits from branch <code>master</code></li><li>c5079543 - [cat-log] Add logic to view workflow file</li><li>4ee3a78c - [cat-log] Add logic to view job file</li><li>7320c24a - Add cat-log for workflow statuses too</li><li>e4c6fe97 - WIP docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12939&start_sha=55e3432182ea30a7de5fef4edc18c4d364683af6)

**Comment by kinow** [comment]
added 2 commits

<ul><li>0e825d16 - [cat-log] Add logic to view job file</li><li>02fd20a1 - [cat-log] Adding docs for cat-log</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12944&start_sha=e4c6fe9798c4322849c65a27f1237d178f07c017)

**Comment by kinow** [comment]
marked the checklist item **write docs** as completed

**Comment by kinow** [comment]
added 2 commits

<ul><li>5f24f1c4 - [cat-log] Add logic to view job file</li><li>5aa0549d - [cat-log] Adding docs for cat-log</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12950&start_sha=02fd20a14bfe0f1249f1b4f0608dc9e9e76ae0dc)

**Comment by kinow** [comment]
marked the checklist item **write tests** as completed

**Comment by kinow** [comment]
marked this merge request as **ready**

**Comment by kinow** [comment]
100% test coverage. Adding a screencast from my IDE here since GitLab is still not able to produce the coverage reports #971 

On the left sidebar red means not covered, and green is covered.

![GIFrecord_2023-03-13_150530](https://earth.bsc.es/gitlab/es/autosubmit/uploads/65b8e7ee3fe65acd61c297c40428f3ac/GIFrecord_2023-03-13_150530.gif)

**Comment by kinow** [comment]
added 1 commit

<ul><li>45a36b30 - [cat-log] add unit tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=12997&start_sha=5aa0549d500281063975846deb1efb1b35d850d8)

**Comment by kinow** [comment]
requested review from @dbeltrankyl

**Comment by kinow** [comment]
@dbeltrankyl assuming the pipeline passes, this should be ready to be reviewed, please :grimacing: 

To test it, run any workflow so that it generates some workflow & jobs logs. Then instead of looking at the files, try `autosubmit cat-log --help` or some commands like

- `autosubmit cat-log a000`  # meaning `cat` the output logs for workflow a000
- `autosubmit cat-log a000 -m t`  # -m is mode, options are `[c]at` to print and exit, and `[t]ail` to call `tail -f`
- `autosubmit cat-log a000 -m t -f s`  # -f is the file, options vary for workflow and jobs, see help, but here s is the status file for the workflow
- `autosubmit cat-log a000_20220401_fc0_PREINI -f o`  # a job log (need full ID) in the mode output


And so it goes :slight_smile: 

Thanks!

-Bruno

**Comment by kinow** [comment]
Hold on, I think I can write a test for this too.

@mcastril , @dbeltrankyl already a big win having coverage on GitLab. On my IDE I saw that my new code below was covered, but completely missed this part of my MR.

Were it not for GitLab displaying coverage, I would have never written tests for it!

**Comment by kinow** [comment]
added 1 commit

<ul><li>8737dbf7 - [cat-log] add unit tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13001&start_sha=45a36b3015b137b66e92cedd0e247a991b78d28f)

**Comment by kinow** [comment]
added 1 commit

<ul><li>d6364df1 - [cat-log] add unit tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13012&start_sha=8737dbf7c9a680114aa2c25d60067ebabe2d6d9f)

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 15, 2023, 24:18

Commented on [autosubmit/autosubmit.py line 628](https://github.com/BSC-ES/autosubmit/compare/365c99198907ce09f6d43787f40eeca66d5c53a4..d6364df1179f1452878c388b185cfbe908a0a640#diff-fb87ba4e4225b357016f4298108bcf0dR628)



It's very nice that we have coverage in GitLab now !

Could you explain we the IDE was showing different things than GitLab? Or maybe I didn't get the difference

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 15, 2023, 10:48

Hello @kinow,

I'm starting the review, 

Maybe it could be helpful also to allow to cat *.cmd? They can be in $EXPID/tmp ( inspect ) or $EXPID/tmp/LOG_a06r ( run ). Maybe the same for %PROJDIR%/scripts

I tried to do it, but it complained: 

```
 [CRITICAL] Invalid cat-log file a06r_REMOTE_SETUP.cmd. Expected one of ['o', 'j', 'e', 's'] [eCode=7011]
More info at https://autosubmit.readthedocs.io/en/master/troubleshooting/error-codes.html
(python39) (base) dbeltran@ bsces107894:~/autosubmit/a06r/tmp/LOG_a06r$ autosubmit cat-log a06r -f a06r_REMOTE_SETUP.cmd
```

Could we also add a flag or a similar command to modify scripts? It could just call vim, nano, gedit, or the default the user configures instead of a cat/tail. I think it may be helpful for developers. We can do this for all the %JOBS.$JOB.FILE% 


What do you think?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 15, 2023, 11:58

Commented on [docs/source/qstartguide/index.rst line 125](https://github.com/BSC-ES/autosubmit/compare/365c99198907ce09f6d43787f40eeca66d5c53a4..d6364df1179f1452878c388b185cfbe908a0a640#diff-86ccf43d9433c872875cb443d434563fR125)



What does autodata:: ? 

The make watch reports:

`autosubmit.autosubmit.Autosubmit.cat_log= <function Autosubmit.cat_log>`

It this the intended result or should it show the --help usage or something like that?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 15, 2023, 12:06

Commented on [test/unit/test_catlog.py line 14](https://github.com/BSC-ES/autosubmit/compare/cb669f8cc2be3e612e6641a8850d3d692441b56a..5a6da290a4c472eda3498f8b13e582cc6f715d0f#diff-d354699954bd0446a9d31ebdfaa03560R14)



The test works for me locally, but it complains about 18 warnings.


This file:

======================= 11 passed, 18 warnings in 0.48s ========================

Total unit folder:

======================= 160 passed, 20 warnings in 1.26s =======================

It doesn't tell me what the warnings are tho.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 15, 2023, 12:07

All is reviewed, great work! all is working as expected just few minor comments

**Comment by kinow** [comment]
>It's very nice that we have coverage in GitLab now !

:tada: 

>Could you explain we the IDE was showing different things than GitLab? Or maybe I didn't get the difference

The IDE and GitLab both show the correct coverage number. The issue was related to I not paying enough attention to the report in the IDE, and also to another issue I wanted to propose in an issue, but I will try to summarize it here too.

The change for `autosubmit cat-log` was in `autosubmit/autosubmit.py`, the application entry point for `autosubmit` commands. That module, `autosubmit.py`, has today 5885 lines, and growing with each new subcommand added, or with modifications to existing commands.

The IDE displayed the change at the bottom of the file, where I had a new function for `cat-log`. But there was some change at the top (some 5000 lines above) that I failed to notice.

There is also an issue with performance having such a long module. If you look at `autosubmit/autosubmit.py`, at the top, there are **heaps** of `import` statements. Python tries to optimize/cache `import`'s, but it depends on how the modules imported are written (also Python flavor [pypy/cpython/ironpython/whatnot] and version). Python docs mention this issue with imports too - https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Import_Statement_Overhead

So for running `autosubmit cat-log`, Python interpreter first needs to load the `autosubmit/autosubmit.py` module, and alllll the Python modules being imported for `autosubmit create`, `autosubmit report`, `autosubmit inspect`, etc. Even though I won't use most of those modules.

-Bruno

**Comment by kinow** [comment]
Created an issue to track that now: https://earth.bsc.es/gitlab/es/autosubmit/-/issues/973

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 15, 2023, 17:08

Commented on [autosubmit/autosubmit.py line 628](https://github.com/BSC-ES/autosubmit/compare/365c99198907ce09f6d43787f40eeca66d5c53a4..d6364df1179f1452878c388b185cfbe908a0a640#diff-fb87ba4e4225b357016f4298108bcf0dR628)



Thank you for the explanation and creating the issue Bruno!

**Comment by kinow** [comment]
I've ran out of energy to work on the review feedback today :disappointed_relieved: But I will try to start working on that tomorrow. But great review, thanks a lot @dbeltrankyl !

**Comment by kinow** [comment]
Great ideas!!! +1 to all your suggestions @dbeltrankyl . Quick question about this one

>They can be in $EXPID/tmp ( inspect ) or $EXPID/tmp/LOG_a06r ( run ). Maybe the same for %PROJDIR%/scripts

What should be the search order? First $EXPID/tmp/LOG_a06r ( run ), and then if not found search in $EXPID/tmp ( inspect )?

>Could we also add a flag or a similar command to modify scripts? It could just call vim, nano, gedit, or the default the user configures instead of a cat/tail. I think it may be helpful for developers. We can do this for all the %JOBS.$JOB.FILE%

I will let `vi` as default, but will check if the `$EDITOR` env var has been set and use that instead when available (maybe the `$VISUAL` too).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 16, 2023, 09:50

> What should be the search order? First, $EXPID/tmp/LOG_a06r ( run ), and then if not found search in $EXPID/tmp ( inspect )?

Not sure. 

1) Run will have the most recent one submitted
2) But inspect would be helpful as well when one is modifying the scripts without submitting them, and maybe there is a cmd on the run folder already ( that would be outdated) 

The safest would be: 

* Default run
* Add a prefix like inspect_$jobname.cmd to have an alias
* Add a prefix like run_$jobname.cmd to have an alias

So the user can choose what to look for exactly.

**Comment by kinow** [comment]
Oh! That was not supposed to be committed yet :sweat_smile: 

I was trying to display the `__doc__` of the `cat-log` function. This way we could document the code for devs, but include it in the documentation too. So that list of parameters & help messages would always stay in sync.

I will remove it as I have another branch where I'm working on that :thumbsup: thanks for spotting this!

**Comment by kinow** [comment]
changed this line in [version 13 of the diff](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13089&start_sha=d6364df1179f1452878c388b185cfbe908a0a640#cc54eb7d219fa47cf8acb699abf650df10979bd5_125_140)

**Comment by kinow** [comment]
added 2 commits

<ul><li>e93f5a52 - [cat-log] Adding docs for cat-log</li><li>6d0cfeb7 - [cat-log] add unit tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13089&start_sha=d6364df1179f1452878c388b185cfbe908a0a640)

**Comment by kinow** [comment]
Modified the docs to contain now a simple output. I've also added a TODO to have a link to more complete docs about the `cat-log` (and other commands). There's an existing issue where I started working on that (can't find it right now).

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/1dda0b4bc8517f709354c3f313b69119/image.png)

Edited the commit with docs and push-forced.

**Comment by kinow** [comment]
How are you running this test?

When I use the IDE with the test option set to Unittests, here's what I get:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/06982a0c0c0f768f1c6880079e684b9c/image.png)

```bash
Launching Nosetest with arguments /home/kinow/Development/python/pycharm-2018.3.4/plugins/python/helpers/pycharm/_jb_nosetest_runner.py /home/kinow/Development/python/workspace/autosubmit/test/unit/test_catlog.py in /home/kinow/Development/python/workspace/autosubmit/test

...........
----------------------------------------------------------------------
Ran 11 tests in 0.035s

OK

Process finished with exit code 0
```

If I use the command used in GitLab CICD pipeline:

```bash
$ python3 -m 'nose' --exclude=regression  --verbosity=3 test/unit --with-coverage --cover-package=autosubmit --cover-inclusive --cover-xml --cover-xml-file=test/coverage.xml
over-xml --cover-xml-file=test/coverage.xml
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.config: INFO: Excluding tests matching ['regression']
nose.plugins.cover: INFO: Coverage report will include only packages: ['autosubmit']
test_read_makes_the_right_method_calls (test.unit.test_basic_config.TestBasicConfig) ... ok
test_update_config_set_the_right_db_path (test.unit.test_basic_config.TestBasicConfig) ... ok
...
...
test_returned_packages_parent_failed_mixed_wrapper (test.unit.test_wrappers.TestWrappers) ... ok
test_returned_packages_section_not_self_dependent (test.unit.test_wrappers.TestWrappers) ... ok

Name                                                                    Stmts   Miss  Cover
-------------------------------------------------------------------------------------------
autosubmit/__init__.py                                                      0      0   100%
autosubmit/autosubmit.py                                                 3650   3137    14%
...
...
autosubmit/statistics/stats_summary.py                                     19     15    21%
autosubmit/statistics/utils.py                                             37     26    30%
-------------------------------------------------------------------------------------------
TOTAL                                                                   13791  10210    26%
----------------------------------------------------------------------
Ran 160 tests in 1.063s

OK
```

I think maybe these warnings might happen if you are using pytest on PyCharm to run the Unittests? If you know how to reproduce and maybe get the output/error logs I can try to fix these warnings.

**Comment by kinow** [comment]
@dbeltrankyl about this one, 

>Maybe it could be helpful also to allow to cat \*.cmd? They can be in $EXPID/tmp ( inspect ) or $EXPID/tmp/LOG_a06r ( run ). Maybe the same for %PROJDIR%/scripts

I think this is already implemented, but only for `.cmd` files on the `'tmp` directory (inspect):

```
$ autosubmit cat-log -f j a000_20220402_fc0_1_1_APPLICATION | head -n 5
Autosubmit is running with 4.0.0b
#!/bin/bash

###############################################################################
#                   APPLICATION a000 EXPERIMENT

```

Will check the other cmd files from run now. Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 21, 2023, 11:27

Hello @kinow , 

Tell me when it requires ( if it is necessary ) a new review

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 21, 2023, 11:28

removed review request for @dbeltrankyl

**Comment by kinow** [comment]
added 20 commits

<ul><li>6d0cfeb7...c87d5422 - 16 commits from branch <code>master</code></li><li>c363b6a9 - [cat-log] Add logic to view workflow file</li><li>750b15d4 - [cat-log] Add logic to view job file</li><li>88b8c6dd - [cat-log] Adding docs for cat-log</li><li>681e5753 - [cat-log] add unit tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13268&start_sha=6d0cfeb733ea532257a1e4a6fdcae0dc9949c8cd)

**Comment by kinow** [comment]
added 1 commit

<ul><li>80a20e38 - Add an `-i` `--`inspect` option to use log files produced by `autosubmit...</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13274&start_sha=681e5753ac9fc37e517174876fa4e3638aee06f2)

**Comment by kinow** [comment]
Done @dbeltrankyl ! Now if you run just `autosubmit cat-log -f j -m c a006_LOCAL_SETUP` it will try to print

- the job/cmd script (`-f j`)
- using `cat` (`-m c`)
- of the `a006_LOCAL_SETUP` job

But if you use `autosubmit cat-log -f j -m c -i a006_LOCAL_SETUP`, then the `-i` (or `--inspect`) will tell the command to look for log files in `tmp/` instead of `tmp/LOG_a000`.

**Comment by kinow** [comment]
I've updated it based on your feedback about files generated by `autosubmit inspect`. I called the new command-line flag `--inspect`, but thinking now, I have not confirmed if it's only `inspect` that generates files in `tmp/`. Let me know if you have a better name for the command. You can take a look at the latest commit to see the diff, and if that makes sense.

I've also updated the docs, checked on coverage, and the CI job passed successfully on GitLab.

Ready for another review whenever you have time, please, @dbeltrankyl .

Thanks!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 23, 2023, 08:51

Tested and working, 

I think that the syntax will be a bit hard to remember for the users, similar to the autosubmit setstatus, 

I'm thinking about doing somewhere in the docs a syntax Quick Reference Card that encapsulates all commands, or some of them in one single doc page so the users can always look at it to have a quick remember/view of the syntax instead of scrolling the whole rst ( or using the search ). They can also use the cli help but I think it is ignored by the majority of users ( in my experience ) 

What do you think?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 23, 2023, 09:02

Hello @kinow , 

* For the remote platforms, only --inspect stores the cmds on tmp. 

* For the local platform, autosubmit run also stores the cmds on there  ( and total_stats , stats , completed ) 

I think inspect is good enough, tho. 

I believe that at some point, we need to change the Local platform to use directly the tmp/LOG_a004 folder. I remember that Miguel Andres noticed this and in some issue we did a comment regarding that

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 23, 2023, 09:36

Commented on [test/unit/test_catlog.py line 14](https://github.com/BSC-ES/autosubmit/compare/cb669f8cc2be3e612e6641a8850d3d692441b56a..5a6da290a4c472eda3498f8b13e582cc6f715d0f#diff-d354699954bd0446a9d31ebdfaa03560R14)



Hello @kinow 

No warnings using nosetests as you said. 

By default, pycharms adds `--no-headers --no-summary -q` to the pytest. 

To disable this default behavior, you have to go to: 

Settings -> advanced settings -> Python -> Pytest: do not add `--no-headers --no-summary -q`

Then the output is as follows: 

```sh
=============================== warnings summary ===============================
../../../../python39/lib/python3.9/site-packages/matplotlib/__init__.py:152
../../../../python39/lib/python3.9/site-packages/matplotlib/__init__.py:152
../../../../python39/lib/python3.9/site-packages/matplotlib/__init__.py:152
../../../../python39/lib/python3.9/site-packages/matplotlib/__init__.py:152
../../../../python39/lib/python3.9/site-packages/matplotlib/__init__.py:152
  /home/dbeltran/python39/lib/python3.9/site-packages/matplotlib/__init__.py:152: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(module.__version__) < minver:

../../../../python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346
../../../../python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346
../../../../python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346
../../../../python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346
../../../../python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346
  /home/dbeltran/python39/lib/python3.9/site-packages/setuptools/_distutils/version.py:346: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    other = LooseVersion(other)

../../../../python39/lib/python3.9/site-packages/autosubmitconfigparser/config/basicconfig.py:84
../../../../python39/lib/python3.9/site-packages/autosubmitconfigparser/config/basicconfig.py:84
test/unit/test_catlog.py::TestJob::test_command_line_help
test/unit/test_catlog.py::TestJob::test_command_line_help
  /home/dbeltran/python39/lib/python3.9/site-packages/autosubmitconfigparser/config/basicconfig.py:84: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
    parser = SafeConfigParser()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================= 11 passed, 14 warnings in 0.55s ========================

```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 23, 2023, 09:39

Commented on [test/unit/test_catlog.py line 14](https://github.com/BSC-ES/autosubmit/compare/cb669f8cc2be3e612e6641a8850d3d692441b56a..5a6da290a4c472eda3498f8b13e582cc6f715d0f#diff-d354699954bd0446a9d31ebdfaa03560R14)



Looking at the warnings, the only one that we can solve is the `parser = Safeconfigparser() for parser = ConfigParser()` ( used for read the autosubmitrc) 

The rest of them are external libs

**Comment by kinow** [comment]
>I think that the syntax will be a bit hard to remember for the users, similar to the autosubmit setstatus,

Could be. One advantage is that anyone familiar with Cylc will already know how to use it with Autosubmit (except for the `--inspect` flag). Likewise, if an Autosubmit operator ever needs to work with Cylc, they will be able to quickly adapt to the `cylc cat-log`.

>I'm thinking about doing somewhere in the docs a syntax Quick Reference Card that encapsulates all commands, or some of them in one single doc page so the users can always look at it to have a quick remember/view of the syntax instead of scrolling the whole rst ( or using the search ). They can also use the cli help but I think it is ignored by the majority of users ( in my experience )
>
>What do you think?

+1

It would be nice if we had this done automatically. That was my initial idea here, in that line with `autodata` that you spotted in the Sphinx code. That way our docs won't get out of sync with the code (i.e. if we modify the code either the docs are automatically updated in the next build, or the build fails until we fix the docs).

**Comment by kinow** [comment]
Lookoing at the warnings above, I think the only one in this change is the `test_command_line_help`. The rest appears to be from libraries as you said.

```
    def test_command_line_help(self):
        args = ['autosubmit', 'cat-log', '--help']
        with patch.object(sys, 'argv', args) as _, io.StringIO() as buf, redirect_stdout(buf):
            with suppress(SystemExit):
                assert Autosubmit.parse_args()
            assert buf
            assert 'View workflow and job logs.' in buf.getvalue()
```

>`parser = Safeconfigparser() for parser = ConfigParser()` ( used for read the autosubmitrc)

Searching the code base, I found it in `autosubmit.py` only. And 1x in `configure()`, and 2x in `configure_dialog()`. Let me know if you want me to replace `Safeconfigparser` by `ConfigParser` here, or create an issue for later?

**Comment by kinow** [comment]
>I think inspect is good enough, tho.

:thumbsup: 

>I believe that at some point, we need to change the Local platform to use directly the tmp/LOG_a004 folder. I remember that Miguel Andres noticed this and in some issue we did a comment regarding that

+1 sounds good to me. Some users may complain, but we can tell them to use `cat-log` to avoid future issues (as we may move more files around :grimacing: ).

BTW, do you know why we have the experiment name twice in the directory structure? i.e. `~/autosubmit/a000/tmp/LOG_a000`. It would be the same if we had `~/autosubmit/a000/tmp/LOG` or `log` (no need to press shift). It is simpler to write scripts/code to automate since you don't need to always concatenate `LOG_` with the expid, and you already know the experiment from the parent-parent `a000` directory?

**Comment by kinow** [comment]
added 66 commits

<ul><li>80a20e38...cb669f8c - 61 commits from branch <code>master</code></li><li>98b9d013 - [cat-log] Add logic to view workflow file</li><li>43063c68 - [cat-log] Add logic to view job file</li><li>da0a85a4 - [cat-log] Adding docs for cat-log</li><li>6bd229f6 - [cat-log] add unit tests</li><li>c37d66ee - Add an `-i` `--`inspect` option to use log files produced by `autosubmit...</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13390&start_sha=80a20e3837162fa81fe1ad90de615f05bddda60c)

**Comment by kinow** [comment]
> dbeltran @dbeltrankyl removed review request for @dbeltrankyl 2 days ago 

p.s. waiting for approval before merging it (not sure if you finished reviewing it or not)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 24, 2023, 09:34

Commented on [test/unit/test_catlog.py line 14](https://github.com/BSC-ES/autosubmit/compare/cb669f8cc2be3e612e6641a8850d3d692441b56a..5a6da290a4c472eda3498f8b13e582cc6f715d0f#diff-d354699954bd0446a9d31ebdfaa03560R14)



You can do it there and merge it if you wish

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 24, 2023, 09:40

Commented on [test/unit/test_catlog.py line 14](https://github.com/BSC-ES/autosubmit/compare/cb669f8cc2be3e612e6641a8850d3d692441b56a..5a6da290a4c472eda3498f8b13e582cc6f715d0f#diff-d354699954bd0446a9d31ebdfaa03560R14)



Mainly due because the configparse was used for all Autosubmit config now it is only the autosubmitrc

**Comment by kinow** [comment]
added 1 commit

<ul><li>c49ae06a - Replace SafeConfigParser (deprecated) by ConfigParser</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13394&start_sha=c37d66ee017afc3dde58b341b62de6539b1f424d)

**Comment by kinow** [comment]
Thanks for confirming it's safe to remove it, @dbeltrankyl! Just did so. Waiting for pipeline to run now, and for all tests to pass before merging it.

**Comment by kinow** [comment]
added 1 commit

<ul><li>5a6da290 - Replace SafeConfigParser (deprecated) by ConfigParser</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/320/diffs?diff_id=13396&start_sha=c49ae06ad3d510d6b3e35698b1269887e5d1da4b)

**Comment by kinow** [comment]
mentioned in commit ad44fa6841bbeee27d2f679e619ccf319e1381c3

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 24, 2023, 10:25

I think it's a very good idea to have a complete guide with all commands and options together, and better if that's automatically.

We could in parallel maintain a cheatsheet with the most used commands.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 24, 2023, 10:37

This is the kind of thing that only someone without our bias can ask. I cannot think of any reason why the $EXPID should be repeated, but for comfortability when you run a recursive grep and don't remember the flag to output the directory... It does not seem to be a good reason.
